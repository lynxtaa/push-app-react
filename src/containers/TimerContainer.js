import React, { useCallback, useState, useRef } from 'react'
import PropTypes from 'prop-types'

import TimerButton from '@components/TimerButton'
import Alert from '@components/Alert'
import ButtonGroup from '@components/ButtonGroup'
import TimerLabel from '@components/TimerLabel'
import useInterval from '../hooks/useInterval'

const Timer = ({ times }) => {
	const [countdown, setCountdown] = useState(times[0].seconds)
	const [counter, setCounter] = useState(times[0].seconds)
	const [timerActive, setTimerActive] = useState(false)

	const start = useRef(0)

	useInterval(
		() => {
			const delta = Date.now() - start.current
			const countdown = counter - Math.floor(delta / 1000)
			if (countdown > 0) {
				setCountdown(countdown)
			} else {
				setCountdown(0)
				setTimerActive(false)
			}
		},
		timerActive ? 1000 : null,
	)

	const handleClick = useCallback(event => {
		const value = Number(event.target.value)
		setCounter(value)
		setCountdown(value)
		setTimerActive(false)
	}, [])

	const resetCountdown = useCallback(() => {
		setCountdown(counter)
		setTimerActive(false)
	}, [counter])

	const runTimer = useCallback(() => {
		start.current = Date.now()
		setTimerActive(true)
	}, [])

	return (
		<div className="mb-3">
			{countdown === 0 && (
				<Alert onClick={resetCountdown}>
					<b>Do next!</b>
				</Alert>
			)}

			<ButtonGroup>
				{times.map(({ seconds, label }) => (
					<TimerButton
						active={counter == seconds}
						key={seconds}
						onClick={handleClick}
						value={seconds}
					>
						{label}
					</TimerButton>
				))}
				<TimerLabel onClick={timerActive ? resetCountdown : runTimer}>
					{countdown}
				</TimerLabel>
			</ButtonGroup>
		</div>
	)
}

Timer.propTypes = {
	times: PropTypes.arrayOf(
		PropTypes.shape({
			seconds: PropTypes.number.isRequired,
			label: PropTypes.string.isRequired,
		}),
	).isRequired,
}

export default Timer
