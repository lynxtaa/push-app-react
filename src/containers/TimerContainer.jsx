import React from 'react'
import PropTypes from 'prop-types'

import TimerButton from '@components/TimerButton'
import Alert from '@components/Alert'
import ButtonGroup from '@components/ButtonGroup'
import TimerLabel from '@components/TimerLabel'

class Timer extends React.Component {
	constructor(props) {
		super(props)

		const [{ seconds }] = this.props.times
		this.state = { counter: seconds, countdown: seconds }

		this.handleClick = this.handleClick.bind(this)
		this.resetCountdown = this.resetCountdown.bind(this)
		this.runTimer = this.runTimer.bind(this)
	}

	handleClick(value) {
		this.stopTimer()
		this.setState({ counter: value, countdown: value })
	}

	resetCountdown() {
		this.setState(prev => ({ countdown: prev.counter }))
	}

	stopTimer() {
		clearInterval(this.interval)
	}

	runTimer() {
		this.interval = setInterval(() => {
			if (this.state.countdown > 0) {
				this.setState(prev => ({ countdown: prev.countdown - 1 }))
			}
			else {
				this.stopTimer()
			}
		}, 1000)
	}

	render() {
		const { counter, countdown } = this.state

		return (
			<div>
				{countdown === 0 && <Alert onClick={this.resetCountdown}><b>Do next!</b></Alert>}

				<ButtonGroup>
					{this.props.times.map(({ seconds, label }) => (
						<TimerButton
							active={counter == seconds}
							key={seconds}
							onClick={this.handleClick}
							value={seconds}
						>
							{label}
						</TimerButton>
					))}
					<TimerLabel onClick={this.runTimer}>{countdown}</TimerLabel>
				</ButtonGroup>
			</div>
		)
	}
}

Timer.propTypes = {
	times: PropTypes.arrayOf(PropTypes.shape({
		seconds: PropTypes.number.isRequired,
		label: PropTypes.string.isRequired,
	})).isRequired,
}

export default Timer

