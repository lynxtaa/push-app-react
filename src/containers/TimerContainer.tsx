import React, { useCallback, useState, useRef } from 'react'

import Alert from 'components/Alert'
import TimerLabel from 'components/TimerLabel'
import useInterval from '../hooks/useInterval'
import { ButtonGroup, Button, Box } from '@chakra-ui/core'

interface Time {
	seconds: number
	label: string
}

interface Props {
	times: Time[]
}

const Timer = ({ times }: Props) => {
	const [countdown, setCountdown] = useState<number>(times[0].seconds)
	const [counter, setCounter] = useState<number>(times[0].seconds)
	const [timerActive, setTimerActive] = useState<boolean>(false)

	const start = useRef<number>(0)

	useInterval(
		() => {
			const delta = Date.now() - start.current
			const countdown = counter - Math.floor(delta / 1000)
			if (countdown > 0) {
				setCountdown(countdown)
			} else {
				setCountdown(0)
				setTimerActive(false)
			}
		},
		timerActive ? 1000 : null,
	)

	const resetCountdown = useCallback(() => {
		setCountdown(counter)
		setTimerActive(false)
	}, [counter])

	const runTimer = useCallback(() => {
		start.current = Date.now()
		setTimerActive(true)
	}, [])

	return (
		<Box mb={4}>
			{countdown === 0 && (
				<Alert onClose={resetCountdown} mb={3}>
					Do next!
				</Alert>
			)}

			<ButtonGroup spacing={0}>
				{times.map(({ seconds, label }, i) => (
					<Button
						isActive={counter === seconds}
						key={seconds}
						borderTopRightRadius="none"
						borderBottomRightRadius="none"
						borderTopLeftRadius={i > 0 ? 'none' : undefined}
						borderBottomLeftRadius={i > 0 ? 'none' : undefined}
						onClick={() => {
							setCounter(seconds)
							setCountdown(seconds)
							setTimerActive(false)
						}}
						variantColor="brand"
					>
						{label}
					</Button>
				))}
				<TimerLabel
					onClick={timerActive ? resetCountdown : runTimer}
					borderTopLeftRadius="none"
					borderBottomLeftRadius="none"
				>
					{countdown}
				</TimerLabel>
			</ButtonGroup>
		</Box>
	)
}

export default Timer
