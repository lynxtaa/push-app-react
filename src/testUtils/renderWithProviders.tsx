import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider, ColorModeProvider } from '@chakra-ui/core'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'

import theme from '../theme'

export default function renderWithProviders(
	ui: React.ReactElement,
	{
		initialEntries,
		initialIndex,
		...rest
	}: Omit<RenderOptions, 'queries'> & {
		initialEntries?: string[]
		initialIndex?: number
	} = {},
): RenderResult {
	return render(ui, {
		...rest,
		wrapper: ({ children }: { children?: React.ReactNode }) => (
			<ThemeProvider theme={theme}>
				<ColorModeProvider>
					<MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
						{children}
					</MemoryRouter>
				</ColorModeProvider>
			</ThemeProvider>
		),
	})
}
