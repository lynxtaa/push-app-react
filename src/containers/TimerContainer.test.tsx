import React from 'react'
import { act, screen } from '@testing-library/react'
import createMockRaf, { MockRaf } from '@react-spring/mock-raf'
import { Globals, FrameLoop } from 'react-spring'

import TimerContainer from './TimerContainer'
import renderWithProviders from '../testUtils/renderWithProviders'
import userEvent from '@testing-library/user-event'

const times = [
	{ label: '1 min', seconds: 60 },
	{ label: '2 min', seconds: 120 },
]

declare const Date: any

const now = Date.now()

Date.now = jest.fn()

let mockRaf: MockRaf

beforeEach(() => {
	Date.now.mockReturnValue(now)

	mockRaf = createMockRaf()

	Globals.assign({
		now: mockRaf.now,
		requestAnimationFrame: mockRaf.raf,
		cancelAnimationFrame: mockRaf.cancel,
		frameLoop: new FrameLoop(),
	})
})

it('renders timer', () => {
	const { container } = renderWithProviders(<TimerContainer times={times} />)
	expect(container.firstChild).toMatchSnapshot()
})

it('runs timer after click', async () => {
	jest.useFakeTimers()

	renderWithProviders(<TimerContainer times={times} />)

	const timerButton = screen.getByText('60')
	userEvent.click(timerButton)

	act(() => {
		Date.now.mockReturnValue(Date.now() + 3000)
		jest.advanceTimersByTime(3000)
	})

	mockRaf.flush()

	expect(timerButton).toHaveTextContent('57')
})

it('shows alert after countdown', async () => {
	jest.useFakeTimers()

	const renderResult = renderWithProviders(<TimerContainer times={times} />)

	const timerButton = renderResult.getByText('60')
	userEvent.click(timerButton)

	act(() => {
		Date.now.mockReturnValue(Date.now() + 60000)
		jest.advanceTimersByTime(60000)
	})

	mockRaf.flush()

	expect(screen.getByText('Do next!')).toBeInTheDocument()
	expect(timerButton).toHaveTextContent('0')
})

it('toggles timer', async () => {
	jest.useFakeTimers()

	renderWithProviders(<TimerContainer times={times} />)

	const timerButton = screen.getByText('60')
	userEvent.click(timerButton)

	act(() => {
		Date.now.mockReturnValue(Date.now() + 3000)
		jest.advanceTimersByTime(3000)
	})

	userEvent.click(timerButton)
	expect(timerButton).toHaveTextContent('60')

	act(() => {
		Date.now.mockReturnValue(Date.now() + 3000)
		jest.advanceTimersByTime(3000)
	})

	expect(timerButton).toHaveTextContent('60')
})

it('stops timer when counter is changed', async () => {
	jest.useFakeTimers()

	renderWithProviders(<TimerContainer times={times} />)

	const timerButton = screen.getByText('60')
	userEvent.click(timerButton)

	act(() => {
		Date.now.mockReturnValue(Date.now() + 3000)
		jest.advanceTimersByTime(3000)
	})

	const anotherCounter = screen.getByText('2 min')
	userEvent.click(anotherCounter)

	act(() => {
		Date.now.mockReturnValue(Date.now() + 3000)
		jest.advanceTimersByTime(3000)
		mockRaf.flush()
	})

	expect(timerButton).toHaveTextContent('120')
})
