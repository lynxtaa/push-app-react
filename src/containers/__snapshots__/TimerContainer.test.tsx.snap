// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders timer 1`] = `
<div
  class="css-acwcvw"
>
  <div
    class="css-1baulvz"
  >
    <button
      class="css-mwy6p6"
      data-active="true"
      type="button"
    >
      1 min
    </button>
    <button
      class="css-13bwko5"
      type="button"
    >
      2 min
    </button>
    <button
      class="css-1wc6wq4"
      type="button"
    >
      60
    </button>
  </div>
</div>
`;
